### ISSUE TITLE
0.9.0: Integrate MARLIN methylation classifier and validate against GSE280090

### ISSUE THREAD

**Issue by xhuljanoshehu** [xhuljanoshehu/ontseq-platform#76]
## Goal

Add a reproducible, fail-closed MARLIN methylation-classification lane to ONTSeq and validate it against the publicly released Nanopore methylation data from **GSE280090** before making any diagnostic or clinical claims.

This issue covers the **downstream methylation-classification path**. It does **not** by itself validate POD5 → Dorado → alignment → MM/ML → modkit.

## Current state

ONTSeq already has:

- a modkit-based methylation lane using pinned **modkit 0.6.4 semantics**;
- bedMethyl/CpG-oriented downstream processing and reviewer visualization;
- in-silico dilution / LoD infrastructure;
- explicit separation of technical execution, analytical validation and clinical reportability;
- fail-closed handling for invalid/missing data.

What is **not yet implemented** is MARLIN inference itself:

```text
CpG / bedMethyl
      ↓
MARLIN feature mapping
      ↓
trained MARLIN model
      ↓
class + score/confidence
      ↓
UNKNOWN / no-call logic
      ↓
ONTSeq report
```

## Reference implementation / publications

Primary reference:

- Steinicke TL, Benfatto S et al. *Rapid epigenomic classification of acute leukemia.* Nature Genetics (2025).
- MARLIN code: https://github.com/hovestadt/MARLIN
- Versioned code: https://doi.org/10.5281/zenodo.15723932
- Trained MARLIN model: https://doi.org/10.5281/zenodo.15565404
- MARLIN training dataset: https://zenodo.org/records/15566584

## External validation dataset

Use **GSE280090** as the first external public validation set:

- GEO series: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE280090
- first target sample: `GSM8587229_AL_001.txt.gz`
- this file is a precomputed CpG/methylation table (BED-/bedMethyl-like downstream representation), not raw POD5/BAM input;
- reference coordinates in the published workflow are hg19/GRCh37;
- GSE280090 provides processed methylation calls but not POD5, unaligned BAM, modBAM or full alignment BAM.

Therefore this dataset can validate:

- file parsing / schema detection;
- coordinate handling;
- CpG feature mapping;
- missing-feature handling;
- MARLIN inference;
- confidence / threshold / UNKNOWN logic;
- reproducibility across runs;
- ONTSeq report rendering.

It **cannot** validate:

- Dorado basecalling;
- MM/ML-tag generation;
- alignment;
- ONTSeq modkit extraction from BAM;
- end-to-end POD5 → MARLIN behavior;
- CNV, SV or fusion calling.

## Proposed architecture

Support two clearly separated methylation entry paths that converge on one canonical normalized methylation representation:

```text
A) modBAM + MM/ML
        ↓
   modkit 0.6.4
        ↓
   normalized CpG dataset

B) precomputed CpG / bedMethyl-compatible input
        ↓
   validated import adapter
        ↓
   normalized CpG dataset

                ↓
        MARLIN feature mapping
                ↓
        pinned trained model
                ↓
      class / score / confidence
                ↓
          UNKNOWN / NO_CALL
                ↓
           ONTSeq report
```

The precomputed-input path must never be presented as evidence that the upstream BAM/modkit path was validated.

### 1. Provenance lock

Record at minimum:

- MARLIN code version / commit or Zenodo version;
- trained-model artifact identifier and checksum;
- reference build;
- expected feature set / feature count;
- ONTSeq version;
- input file checksum;
- preprocessing/import adapter version.

### 2. Precomputed methylation import

Implement an explicit adapter for published CpG/bedMethyl-like files.

Requirements:

- gzip input supported;
- schema validated before analysis;
- coordinate convention explicitly checked;
- build must be explicit, never guessed silently;
- malformed numeric values fail closed;
- duplicate CpGs handled deterministically;
- missing CpGs represented according to MARLIN's published inference contract;
- imported data marked as `PRECOMPUTED_METHYLATION`, distinct from `MODKIT_DERIVED`.

### 3. MARLIN inference adapter

- use the published trained model rather than retraining for the first integration;
- pin model and preprocessing artifacts;
- reproduce published class labels and scoring semantics;
- preserve an explicit UNKNOWN / insufficient-confidence state;
- do not convert low-confidence or unsupported samples into a forced diagnosis.

#### Phase A — single-sample smoke / reproducibility

Start with:

`GSM8587229_AL_001.txt.gz`

Require:

- deterministic parse;
- deterministic feature vector;
- deterministic MARLIN output;
- stable class/score across repeated runs;
- provenance written to normalized JSON/report.

#### Phase B — GSE280090 cohort

Run the available public Nanopore methylation samples through the same locked adapter/model.

Predefine comparison outputs:

- predicted class;
- confidence / score;
- expected published class where available;
- concordant / discordant / unknown;
- missing-feature burden;
- runtime;
- parser/inference failures;
- no-call rate.

Do not tune thresholds post hoc to maximize concordance.

#### Phase C — bridge to native ONTSeq methylation path

Once a real authorized modBAM/raw path is available:

```text
same specimen
published/precomputed CpG path
            vs
ONTSeq BAM → modkit 0.6.4 → CpG path
```

Compare the normalized CpG representation and resulting MARLIN inference.

This is the required bridge before claiming that the native ONTSeq methylation pipeline reproduces MARLIN end to end.

## Acceptance criteria

- [ ] documented and pinned MARLIN code/model artifacts
- [ ] model checksum recorded
- [ ] explicit GRCh37/hg19 handling implemented
- [ ] gzip precomputed-CpG import supported
- [ ] strict schema/coordinate validation
- [ ] canonical normalized methylation representation defined
- [ ] source provenance distinguishes `PRECOMPUTED_METHYLATION` from `MODKIT_DERIVED`
- [ ] MARLIN inference integrated behind a stable adapter
- [ ] class, score/confidence and UNKNOWN state represented in normalized result schema
- [ ] `GSM8587229_AL_001.txt.gz` produces a deterministic reproducible result
- [ ] GSE280090 cohort validation report generated with concordance/no-call metrics
- [ ] no post-hoc threshold tuning
- [ ] no clinical-reportable status introduced
- [ ] upstream POD5/Dorado/MM-ML/modkit validation remains a separate gate
- [ ] regression fixtures added for parser, feature mapping and inference
- [ ] validation outcome linked into the ONTSeq 0.9.0 validation audit

## Scientific boundary

Passing this issue would support the statement:

> ONTSeq can reproducibly import published CpG-level Nanopore methylation data and reproduce a locked MARLIN downstream classification workflow.

It would **not** support the stronger statement:

> ONTSeq has clinically validated end-to-end Nanopore methylation classification from raw signal.

That stronger claim requires separate validation of POD5/Dorado, alignment, MM/ML, modkit extraction, analytical LoD/reproducibility and intended-use clinical cohorts.

## Relation to ONTSeq 0.9.0

Treat this as a dedicated methylation-classification lane within the 0.9.0 validation-audit work. Keep the same project rule used elsewhere: **technical PASS ≠ analytical validity ≠ clinical validity**.

**Comment by xhuljanoshehu** [comment]
### Pre-registered AL_001 comparison target — before ONTSeq MARLIN execution

For the first external downstream validation sample (`GSM8587229`, `AL_001`), the comparison target is fixed **before** inspecting any ONTSeq MARLIN result:

- Public input: `GSM8587229_AL_001.txt.gz` from GEO `GSE280090`
- Published processing/build: hg19 / GRCh37; GEO states Dorado v0.3.2 and modkit v0.1.13 were used to derive the processed methylation calls.
- Expected dominant current methylation class from Steinicke/Benfatto et al.: **B-ALL Ph/Ph-like**
- Published final score: approximately **0.69**
- With the fixed published ONTSeq/MARLIN confidence threshold of `0.8`, the pre-registered expected decision is therefore **`COMPLETED + UNKNOWN`**, not `HIGH_CONFIDENCE` and not `NO_CALL`, assuming the downloaded restricted CpG file matches the specified adapter contract and contains observed model features.

Primary references:
- Nature Genetics: https://doi.org/10.1038/s41588-025-02321-z
- GEO sample: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM8587229

This registration does **not** assume an exact ONTSeq floating-point score beyond the publication-supported ~0.69 value and does not authorize threshold or preprocessing changes after seeing the ONTSeq outcome. It applies only to the processed-CpG downstream reproduction gate; it is not evidence for POD5/Dorado/alignment/MM-ML/modkit equivalence.

**Comment by xhuljanoshehu** [comment]
### Audit finding — native MARLIN preprocessing is not yet toolchain-equivalent

**Pinned sources inspected:** `main=a81628014506dbc94b1ed77c7b9c8bc4bc982cf0`, ONTSeq package `0.8.2`; upstream MARLIN commit `442aa603415a54f62e7367794f9a31c6bc20fc2d`; current ONTSeq modkit `v0.6.4` tag `cd85862f71d3bfc289f12adc1052a2e574c95e0f`; historical MARLIN/public-data modkit `v0.1.13` commit `543d4d20c1df9996bb4ecb2c5d766ae2e82f6d95`.

This is a **refinement of the existing M-013 preprocessing-drift finding**, not a new classifier-error claim.

The current ONTSeq native lane runs roughly:

```text
modkit pileup ... --modified-bases 5mC 5hmC --combine-mods --cpg \
  --include-bed <model-probes> --filter-threshold 0.8
```

The pinned MARLIN realtime script instead runs:

```text
modkit pileup --ref <hg19.fa> --include-bed <marlin_v1.probes_hg19.bed> \
  -t 10 --combine-mods --only-tabs <bam> <pileup>
```

The downstream count aggregation is aligned in principle: MARLIN `2_process_pileup.R` computes `sum(cov_mod)/sum(cov_valid)` per `probe_id`, matching ONTSeq weighted pooling. The important divergence is upstream of that aggregation.

#### Version bridge is part of the contract

The publication/GEO processing domain used **Dorado v0.3.2 and modkit v0.1.13**. Exact v0.1.13 source accepts `--include-bed` without requiring `--motif` or `--modified-bases`; if no explicit `--filter-threshold` is supplied it calls the data-derived threshold path (`get_threshold_from_options`, default filter percentile 0.1). That is compatible with the form of the pinned MARLIN command.

Current modkit **v0.6.4 is not CLI-equivalent to that historical toolchain**: its source explicitly fails when `--include-bed` is supplied without `--motif` or `--modified-bases`. Therefore the historical MARLIN command must **not** be treated as an exact A/B lane simply by running the same argv against v0.6.4. ONTSeq's added `--modified-bases 5mC 5hmC`/`--cpg` flags partly satisfy a modern CLI contract, but they also select a modern processing path and require qualification rather than being assumed semantically neutral.

The explicit ONTSeq `--filter-threshold 0.8` remains the highest-priority analytic discrepancy. modkit v0.6.4 documents that without an explicit threshold it estimates a pass threshold from the input modification-probability distribution (default 10th percentile); calls below the pass threshold leave `N_valid`. Thus a fixed 0.8 can alter `N_valid`, `N_mod`, pooled beta and the MARLIN +/-1 feature around beta=0.5. This **base-modification probability threshold is analytically distinct from MARLIN's model-score threshold of 0.8**; the latter does not justify the former.

A patient-free numeric mechanism check illustrates possible directionality without claiming a real-sample effect. With one modified call at 0.95 plus two canonical calls at 0.75, a hypothetical pass threshold 0.70 yields beta=1/3 -> feature -1, while 0.80 retains only the modified call and yields beta=1 -> +1. Conversely, two modified calls at 0.75 plus one canonical call at 0.95 change from beta=2/3 -> +1 to beta=0 -> -1. This is a mechanism proof only; actual divergence must be measured with real modkit outputs.

### Required validation contract — MARLIN historical-vs-current preprocessing harness

Do **not** silently switch production preprocessing. Use synthetic/public non-private inputs and separate the comparison into versioned lanes:

1. **HISTORICAL:** pinned/public MARLIN domain using modkit `v0.1.13` plus the locked MARLIN hg19/probe resources and historical-compatible argv semantics.
2. **CURRENT:** exact ONTSeq native lane with the currently qualified modkit `v0.6.4` binary/identity and current argv.
3. **FACTOR-ISOLATION:** within v0.6.4, vary one factor at a time (`--filter-threshold 0.8`, CpG restriction, explicit modification selection) only to identify mechanism. This lane must not be labelled an exact reproduction of the historical command.

For every lane record tool version/commit/binary identity, reference/probe hashes, effective/estimated threshold, per-position `N_valid/N_mod`, per-probe beta, observed model-feature count, number/fraction of beta>=0.5 feature flips, feature-vector SHA-256, 42-score vector (including max absolute score delta), top class/ranking, UNKNOWN/NO_CALL state, runtime and RAM. Stratify external validation by coverage/sparsity, chemistry, basecaller/modbase model and batch/domain. Processed-CpG reproduction alone is not evidence for POD5/Dorado/MM-ML/modkit equivalence.

**Acceptance:** either demonstrate equivalence within predeclared tolerances in the intended assay domains, or explicitly version the native preprocessing contract and document the measured delta. Until then, documentation should not state unqualified reproduction of the pinned MARLIN pileup semantics.

No patient data, production threshold change, code patch, merge or release is introduced by this audit comment.
